File: DynamicProgramming/2D_DP/GridDP/MaximalSquare.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int maxSqr = 0;
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>>dp(m, vector<int>(n, 0));
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (r == 0 || c == 0) {
                    if (matrix[r][c] == '1') {
                        dp[r][c] = 1;
                        maxSqr = max(maxSqr, dp[r][c]);
                    }
                }
                else if (matrix[r][c] == '1' && matrix[r-1][c-1] == '1') {
                    dp[r][c] = dp[r-1][c-1] + 1;
                    maxSqr = max(maxSqr, dp[r][c]);
                }
            }
        }
        return maxSqr;
    }
};

int main() {
    Solution mySolu = Solution();
    vector<vector<char>> n = {
    {'1', '0', '1', '0', '0'},
    {'1', '0', '1', '1', '1'},
    {'1', '1', '1', '1', '1'},
    {'1', '0', '0', '1', '0'}
};
    cout << mySolu.maximalSquare(n);
}

File: DynamicProgramming/2D_DP/GridDP/MaximalSquare.py
class Solution:
    def maximalSquare(self, matrix: list[list[str]]) -> int:
        maxSqr = 0
        m = len(matrix)
        n = len(matrix[0])
        dp = [[0 for i in range(n)] for i in range(m)]
        for r in range(m):
            for c in range(n):
                if r == 0 or c == 0:
                    if matrix[r][c] == '1':
                        dp[r][c] = 1
                        maxSqr = max(maxSqr, dp[r][c])
                    continue
                elif matrix[r][c] == '1':
                    dp[r][c] = 1
                    maxSqr = max(maxSqr, dp[r][c])
                if r != 0 and c != 0 and matrix[r][c] == '1' and matrix[r-1][c-1] == '1' and matrix[r-1][c] == '1' and matrix[r][c-1] == '1':
                    dp[r][c] = min(dp[r-1][c-1], dp[r][c-1], dp[r-1][c]) + 1
                    maxSqr = max(maxSqr, dp[r][c])
        return maxSqr**2
    
mySolu = Solution()
print(mySolu.maximalSquare([["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]))